improver/blending/spatial_weights.py

```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
+
+        """
+        is_rescaled = iris.cube.CubeList()
+        rescaled_weights = iris.cube.CubeList()
+        for weights_slice in weights.slices_over(self.blend_coord):
+            weights_nonzero = np.where(weights_slice.data > 0, 1, 0)
+            if np.all(weights_nonzero == 1):
+                # if there are no masked points in this slice, keep current weights
+                # and mark as unchanged (not rescaled)
+                rescaled_weights.append(weights_slice)
+                is_rescaled.append(
+                    weights_slice.copy(data=np.zeros(weights_slice.shape, dtype=int))
+                )
+            else:
+                weights_orig = weights_slice.data.copy()
+
+                # calculate the distance to the nearest invalid point, in grid squares,
+                # for each point on the grid
+                distance = distance_transform_edt(weights_nonzero)
+
+                # calculate a 0-1 scaling factor based on the distance from the
+                # nearest invalid data point, which scales between 1 at the fuzzy length
+                # towards 0 for points closest to the edge of the mask
+                fuzzy_factor = rescale(
+                    distance, data_range=[0.0, self.fuzzy_length], clip=True
+                )
+
+                # multiply existing weights by fuzzy scaling factor
+                rescaled_weights_data = np.multiply(
+                    weights_slice.data, fuzzy_factor
+                ).astype(FLOAT_DTYPE)
+                rescaled_weights.append(weights_slice.copy(data=rescaled_weights_data))
+
+                # identify spatial points where weights have been rescaled
+                is_rescaled_data = np.where(
+                    rescaled_weights_data != weights_orig, 1, 0
+                ).astype(int)
+                is_rescaled.append(weights_slice.copy(data=is_rescaled_data))
+
+        weights = rescaled_weights.merge_cube()
+        rescaled = is_rescaled.merge_cube()
+
+        return weights, rescaled
+
+    def _rescale_unmasked_weights(self, weights, is_rescaled):
+        """Increase weights of unmasked slices at locations where masked slices
+        have been smoothed, so that the sum of weights over self.blend_coord is
+        re-normalised (sums to 1) at each point and the relative weightings of
+        multiple unmasked slices are preserved.  Modifies weights cube in place.
+
+        Args:
+            weights (iris.cube.Cube):
+                Cube of weights to which fuzzy smoothing has been applied to any
+                masked slices
+            is_rescaled (iris.cube.Cube):
+                Cube matching weights.shape, with value of 1 where masked weights
+                have been rescaled, and 0 where they are unchanged.
+        """
+        rescaled_data = np.multiply(weights.data, is_rescaled.data)
+        unscaled_data = np.multiply(weights.data, (1 - is_rescaled.data))
```

Comment:
If `is_rescaled` is a `bool` array, you can use the bitwise operator here `~` instead of `1-`

Reply:
Will that multiply correctly?  The point is I want zeros vs original values.  I suppose I could use a `np.where` with a boolean array.

Reply:
It should do :-)

Reply:
OK, will give this a go.